File: Anul II/Medii de proiectare si programare/Laborator/lab03/Java/src/main/java/com/ubb/repository/UserDBRepository.java
package com.ubb.repository;

import com.ubb.model.User;
import com.ubb.utils.JDBCUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Optional;
import java.util.Properties;

public class UserDBRepository implements IUserRepository {
    private final JDBCUtils jdbcUtils;
    private static final Logger logger = LogManager.getLogger();

    public UserDBRepository(Properties properties) {
        logger.info("Initializing UserDBRepository with properties: {} ", properties);
        jdbcUtils = new JDBCUtils(properties);
    }

    @Override
    public Optional<User> getByUsername(String username) {
        logger.traceEntry("getByUsername with task {} ", username);
        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "SELECT * FROM users WHERE username=?")) {
            preparedStatement.setString(1, username);
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                if (resultSet.next()) {
                    User user = extract(resultSet);
                    logger.traceExit(user);
                    return Optional.of(user);
                }
            }
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
        logger.traceExit("null");
        return Optional.empty();
    }


    @Override
    public Optional<User> get(Long id) {
        logger.traceEntry();
        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "SELECT * FROM users WHERE id=?")) {
            preparedStatement.setLong(1, id);
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                if (resultSet.next()) {
                    User user = extract(resultSet);
                    logger.traceExit(user);
                    return Optional.of(user);
                }
            }
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
        logger.traceExit("null");
        return Optional.empty();
    }

    @Override
    public Collection<User> read() {
        logger.traceEntry();
        Connection connection = jdbcUtils.getConnection();
        Collection<User> users = new ArrayList<>();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "SELECT * FROM users")) {
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    User user = extract(resultSet);
                    users.add(user);
                }
            }
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
        logger.traceExit(users);
        return users;
    }

    @Override
    public void add(User newEntity) {
        logger.traceEntry("saving task {} ", newEntity);
        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "INSERT INTO users (username, fullName, password) VALUES (?,?,?)")) {
            preparedStatement.setString(1, newEntity.getUsername());
            preparedStatement.setString(2, newEntity.getFullName());
            preparedStatement.setString(3, newEntity.getPasswordToken());
            int result = preparedStatement.executeUpdate();
            logger.traceExit("Saved {} instances", result);
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
    }

    @Override
    public void delete(Long id) {
    }

    @Override
    public void update(User updatedEntity) {
        logger.traceEntry("updating with task {} ", updatedEntity);
        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "UPDATE users SET fullName=? WHERE id=?")) {
            preparedStatement.setString(1, updatedEntity.getFullName());
            preparedStatement.setLong(2, updatedEntity.getID());
            int result = preparedStatement.executeUpdate();
            logger.traceExit("Updated {} instances", result);
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
    }

    private User extract(ResultSet resultSet) throws SQLException {
        Long id = resultSet.getLong("id");
        String username = resultSet.getString("username");
        String fullName = resultSet.getString("fullName");
        String passwordToken = resultSet.getString("password");

        User user = new User(username, fullName);
        user.setID(id);
        user.setPasswordToken(passwordToken);
        return user;
    }
}

File: Anul II/Medii de proiectare si programare/Laborator/lab03/C#/Ubb.BikeContest.Repository/src/UserDbRepository.cs
using System.Collections;
using System.Data;
using log4net;
using Ubb.BikeContest.Model;

namespace Ubb.BikeContest.Repository;

public class UserDbRepository : IUserRepository
{
    private static readonly ILog Log = LogManager.GetLogger("UserDbRepository");
    private IDictionary<String, string> _props;

    public UserDbRepository(IDictionary<String, string> props)
    {
        Log.Info("Creating UserDbRepository ");
        _props = props;
    }

    public User Get(long id)
    {
        Log.InfoFormat("Entering Get with value {0}", id);
        IDbConnection connection = DBUtils.GetConnection(_props);

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "SELECT id,username,fullName,password FROM users WHERE id=@id";
            IDbDataParameter paramId = command.CreateParameter();
            paramId.ParameterName = "@id";
            paramId.Value = id;
            command.Parameters.Add(paramId);

            using (var dataReader = command.ExecuteReader())
            {
                if (dataReader.Read())
                {
                    User user = Extract(dataReader);
                    Log.InfoFormat("Exiting Get with value {0}", user);
                    return user;
                }
            }
        }
        Log.InfoFormat("Exiting Get with value {0}", null);
        return null;
    }

    public IEnumerable Read()
    {
        Log.InfoFormat("Entering Read");
        IDbConnection connection = DBUtils.GetConnection(_props);
        IList<User> users = new List<User>();

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "SELECT id,username,fullName,password FROM users";
            using (var dataReader = command.ExecuteReader())
            {
                while (dataReader.Read())
                {
                    User user = Extract(dataReader);
                    users.Add(user);
                }
            }
        }
        Log.InfoFormat("Exiting Read with value {0}", users);
        return users;
    }

    public void Add(User newEntity)
    {
        Log.InfoFormat("Entering Add with value {0}", newEntity);
        IDbConnection connection = DBUtils.GetConnection(_props);

        using (var command = connection.CreateCommand())
        {
            command.CommandText = 
                "INSERT INTO users(username, fullName, password) " +
                "VALUES(@username, @fullName, @password);";
            var username = command.CreateParameter();
            username.ParameterName = "@username";
            username.Value = newEntity.Username;
            command.Parameters.Add(username);
            
            var fullName = command.CreateParameter();
            fullName.ParameterName = "@fullName";
            fullName.Value = newEntity.FullName;
            command.Parameters.Add(fullName);
            
            var password = command.CreateParameter();
            password.ParameterName = "@password";
            password.Value = newEntity.Password;
            command.Parameters.Add(password);

            var result = command.ExecuteNonQuery();
            Log.InfoFormat("Added {0} entities", result);
        }
    }

    public void Delete(long id)
    {
        Log.InfoFormat("Entering Delete with value {0}", id);
        IDbConnection connection = DBUtils.GetConnection(_props);
        using (var command = connection.CreateCommand())
        {
            command.CommandText = "DELETE FROM users WHERE id=@id";
            var paramId = command.CreateParameter();
            paramId.ParameterName = "@id";
            paramId.Value = id;
            command.Parameters.Add(paramId);
            var result = command.ExecuteNonQuery();
            Log.InfoFormat("Deleted {0} entities", result);
        }
    }

    public void Update(User updatedEntity)
    {
        Log.InfoFormat("Entering Update with value {0}", updatedEntity);
        IDbConnection connection = DBUtils.GetConnection(_props);

        using (var command = connection.CreateCommand())
        {
            command.CommandText = 
                "UPDATE users SET username=@username, fullName=@fullName, password=@password WHERE id=@id";
            var username = command.CreateParameter();
            username.ParameterName = "@username";
            username.Value = updatedEntity.Username;
            command.Parameters.Add(username);
            
            var fullName = command.CreateParameter();
            fullName.ParameterName = "@fullName";
            fullName.Value = updatedEntity.FullName;
            command.Parameters.Add(fullName);
            
            var password = command.CreateParameter();
            password.ParameterName = "@password";
            password.Value = updatedEntity.Password;
            command.Parameters.Add(password);

            var id = command.CreateParameter();
            id.ParameterName = "@id";
            id.Value = updatedEntity.Id;
            command.Parameters.Add(id);

            var result = command.ExecuteNonQuery();
            Log.InfoFormat("Updated {0} entities", result);
        }
    }

    public User GetByUsername(string username)
    {
        Log.InfoFormat("Entering GetByUsername with value {0}", username);
        IDbConnection connection = DBUtils.GetConnection(_props);

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "SELECT id,username,fullName,password FROM users WHERE username=@username";
            var paramUsername = command.CreateParameter();
            paramUsername.ParameterName = "@username";
            paramUsername.Value = username;
            command.Parameters.Add(paramUsername);

            using (var dataReader = command.ExecuteReader())
            {
                if (dataReader.Read())
                {
                    User user = Extract(dataReader);
                    Log.InfoFormat("Exiting GetByUsername with value {0}", user);
                    return user;
                }
            }
        }
        Log.InfoFormat("Exiting GetByUsername with value {0}", null);
        return null;
    }
    
    private User Extract(IDataReader dataReader)
    {
        long id = dataReader.GetInt64(0);
        String username = dataReader.GetString(1);
        String fullName = dataReader.GetString(2);
        String password = dataReader.GetString(3);

        User user = new User(username, fullName, password);
        user.Id = id;
        return user;
    }
}